### PR TITLE
Introduces AutoGenBench

### LINKED ISSUE #995: Rename `run_scenarios` to `autogenbench`, and create python cli package

### PR DESCRIPTION
**Pull request by afourney**
## Why are these changes needed?

This PR introduced AutoGenBench -- a tool for running common benchmarks, and other templated tests, with the AutoGen framework. It replaces the former "testbed" tool.

For full details, see the [AutoGenBench README.md](https://github.com/microsoft/autogen/blob/autogenbench/samples/tools/testbed/README.md)

TL;DR:
It is a (pip) installable module that handles benchmarking and evaluation tasks. An example session might resemble the following:

```
autogenbench clone --branch autogenbench HumanEval
cd HumanEval
autogenbench run Tasks/r_human_eval_two_agents.jsonl
autogenbench tabulate results/r_human_eval_two_agents
```

Where:
- `autogenbench clone --branch autogenbranch HumanEval` downloads and expands the HumanEval benchmark scenario (from the autogenbench branch here).
- `autogenbench run Tasks/r_human_eval_two_agents.jsonl` runs the tasks defined in `Tasks/r_human_eval_two_agents.jsonl`
- `autogenbench tablue results/r_human_eval_two_agents` tabulates the results of the run

**Note:** You must include `--branch autogenbench` in the clone command when testing, otherwise it will look to the main branch and not find anything.

## Related issue number

Closes #995 , #987 , #996

Supersedes #997

## Checks

- [ ] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

**Comment by qingyun-wu** [comment]
@afourney code formatting issues were fixed by running `pre-commit run --all-files` from my side. Also fixed some minor wording issues.

**Comment by afourney** [comment]
> @afourney code formatting issues were fixed by running `pre-commit run --all-files` from my side. Also fixed some minor wording issues.

Thanks. I'll re-install pre-commit hooks on my end, and hope the issues don't come back.

**Comment by Josephrp** [comment]
i'll try this out , seems really interesting and useful

**Comment by gagb** [comment]
suggest renaming tag `testbed` to `autogenbench` post this merge for improved discovery of tags.

File: samples/tools/testbed/README.md [hunk 1]
```diff
@@ -1,239 +1,171 @@
...
 ```
 
-## Results
-
-By default, the Testbed stores results in a folder hierarchy with the following template:
-
-``./results/[scenario]/[instance_id]/[repetition]``
-
-For example, consider the following folders:
+## Docker Requirement
```
Comment (gagb) [inline samples/tools/testbed/README.md:44]: Suggest moving this up before you ask people to `pip install``

File: samples/tools/testbed/README.md [hunk 2]
```diff
@@ -1,239 +1,171 @@
...
-# Convert tasks
-python utils/prepare_autogpt.py
+- *timestamp.txt*: records the date and time of the run, along with the version of the pyautogen library installed
+- *console_log.txt*: all console output produced by Docker when running autogen. Read this like you would a regular console.
+- *[agent]_messages.json*: for each Agent, a log of their messages dictionaries
+- *./coding*: A directory containing all code written by Autogen, and all artifacts produced by that code.
 
-# Run all the scenarios with GPT-4
-python run_scenarios.py scenarios/AutoGPT/autogpt_twoagent_gpt4.jsonl
+## Contributing or defining new tasks or benchmarks
```
Comment (gagb) [inline samples/tools/testbed/README.md:169]:
```suggestion
## Contributing or Defining New Tasks or Benchmarks
```

**Comment by gagb** [comment]
Nice PR! Left very minor feedback. Otherwise looks fantastic!

**Comment by afourney** [comment]
I added a draft of a blog post. @qingyun-wu to review.

**Comment by afourney** [comment]
> suggest renaming tag `testbed` to `autogenbench` post this merge for improved discovery of tags.

I'm going to give this a try, but it's going to be a little tough -- the autogenbench clone command works by linking to the repo, and the path is part of the current deployment so existing PyPI installs will at least temporarily break.

**Comment by afourney** [comment]
NOTE: Once this is merged, and before deleting the "autogenbench" branch, we will need to follow-up with another super-quick PR to point "autogenbench clone" to the main branch. I can't do that now because the files don't exist in that branch yet, so manual testing will break. I suspect this will be a one-time issue.

**Review by gagb — APPROVED** [review]

**Comment by gagb** [comment]
> NOTE: Once this is merged, and before deleting the "autogenbench" branch, we will need to follow-up with another super-quick PR to point "autogenbench clone" to the main branch. I can't do that now because the files don't exist in that branch yet, so manual testing will break. I suspect this will be a one-time issue.

Can you fix in the code? First try via main and if the file doesn't exists fall back to the other branch?

**Comment by afourney** [comment]
> > NOTE: Once this is merged, and before deleting the "autogenbench" branch, we will need to follow-up with another super-quick PR to point "autogenbench clone" to the main branch. I can't do that now because the files don't exist in that branch yet, so manual testing will break. I suspect this will be a one-time issue.
> 
> Can you fix in the code? First try via main and if the file doesn't exists fall back to the other branch?

I added a --branch switch to the clone command.

Now you can do:

```
autogenbench clone --branch autogenbench HumanEval   
```

to clone from this branch. Otherwise it defaults to main.

**Comment by afourney** [comment]
@ekzhu @qingyun-wu @victordibia @sonichi @rickyloynd-microsoft  

Folks, I'm trying to get more eyeballs on this so that we can get this merged this week. Thanks @gagb for the earlier review.

If testing, be sure to use the "--branch autogenbench" parameter with the clone command since the files don't yet exist in main.

**Comment by qingyun-wu** [comment]
This might be minor (or not): folder names should all use lowercase and underscores following Python convention. For special terms e.g., benchmark names such as AutoGPT, GAIA, it may make sense to keep them as is, but other general names such as folders `Scripts` `Templates`, `Tasks`, I think it better to use all lowercase.

**Comment by afourney** [comment]
> Thanks, @afourney, for the tremendous effort and fantastic work! This might be minor (or not): folder names should all use lowercase and underscores following Python convention. For special terms e.g., benchmark names such as AutoGPT, GAIA, it may make sense to keep them as is, but other general names such as folders `Scripts` `Templates`, `Tasks`, I think it better to use all lowercase.

Yeah, that's unfortunately both minor (in terms of complexity), and also a lot of work to fix (in terms of files to change, and testing to do). Can this be handled in a PR after?

**Review by qingyun-wu — APPROVED** [review]
I am fine merging this PR.

File: samples/tools/autogenbench/pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,49 @@
+[build-system]
+requires = ["setuptools", "setuptools-scm"]
+build-backend = "setuptools.build_meta"
+
+[project]
+name = "autogenbench"
+authors = [
+  { name="Autogen Team", email="autogen@microsoft.com" },
```
Comment (qingyun-wu) [inline samples/tools/autogenbench/pyproject.toml:8]: The email address we use for the AutoGen team is `auto-gen@outlook.com`. Should we use this one?

Comment (afourney) [reply samples/tools/autogenbench/pyproject.toml:8]: Ok, will do.

**Pull request merged**